### PR TITLE
fix(crawl): resilient per-country crawl — retry + carry-forward last-good

### LINKED ISSUE #54: fix(crawl): resilient per-country crawl — retry + carry-forward last-good
## Parent

[\[PRD\] Phase 5: V1 launch polish](https://github.com/taewanu/sounds-abroad/issues/51)

## Problem

The daily crawl rebuilds the charts payload from scratch each run. When Apple RSS transiently fails for a single country (HTTP 500, or a momentary ≠25-item feed), `crawlCountry` writes `{valid:false, tracks:[]}`, overwriting the previous good data. With no retry and no carry-forward, that country ships empty until the next fully-successful crawl. The crawl runs 4×/day and Apple tends to fail a *different* country most runs, so production almost always has ~1 empty country — breaking the core 40-country promise.

Evidence (6/2):
- 6/01 22:17 crawl: `sg` (Singapore) shipped empty.
- 6/02 05:54 crawl: `gb` (United Kingdom) shipped empty — log: `[crawl gb] RSS failed: Apple RSS returned 500 Internal Server Error`.
- Both feeds returned a healthy 25 minutes later — confirming transient, not a data gap.

## Root cause

- `scripts/crawl/run.ts` — `crawlCountry` catches `AppleRssError` and returns an empty `valid:false` country; `crawlAll` builds a fresh map each run with no reference to the prior payload → no carry-forward.
- `scripts/crawl/apple-rss.ts` — no retry on the RSS fetch, so a single transient 500 is fatal for that country.
- (secondary) `RssResponseSchema` requires exactly 25 results; the app schema accepts ≤25.

## Fix (scope: 1 + 2 + 3)

1. **Retry with backoff** on the RSS fetch (2–3 attempts, short backoff) — absorbs transient 500s / blips.
2. **Carry-forward last-good** — when a country still fails after retries, reuse its data from the currently published payload instead of writing it empty. Read the existing blob at crawl start; merge per country.
3. **valid_count guard** — surface `valid_count < 40` as a Sentry warning (currently ships silently under a green run).

## Acceptance criteria

- [x] RSS fetch retries transient failures before giving up
- [x] A country that fails all retries keeps its previously published data (not emptied)
- [x] A fresh country with no prior data that fails everything still yields `valid:false` without crashing
- [x] `valid_count < 40` produces a Sentry warning
- [x] Unit tests cover: retry success, retry exhaustion → carry-forward, carry-forward with no prior data, valid_count guard
- [x] Local CI matrix green (format, lint, typecheck, test, build)

## Out of scope

- Loosening `.length(25)` → fold in only if trivial, else V2
- Per-track iTunes lookup resilience (already degrades to `previewUrl: null`)
- Alerting/paging beyond a Sentry warning → V2

### PR DESCRIPTION
**Pull request by taewanu**
## What

Make the daily chart crawl resilient to transient Apple RSS failures, so a single Apple 500 (or a momentary incomplete feed) no longer ships a country empty until the next clean crawl.

## Why

Auditing the live payload found Singapore empty (`valid:false`, 0 tracks); the next crawl emptied the UK instead — log: `[crawl gb] RSS failed: Apple RSS returned 500 Internal Server Error`. Both feeds were healthy minutes later. The crawler rebuilds the payload from scratch each run with no retry and no carry-forward, so roughly one country is empty in production at any time, the victim rotating each crawl — breaking the 40-country promise.

## Changes

- **Retry** — new `withRetry` retries transient RSS errors with exponential backoff (gated to `AppleRssError`).
- **Carry-forward** — `crawlAll` reuses the last published valid entry when a country still fails after retries; a fresh country with no prior data and a total failure stays `valid:false` without crashing.
- **Degraded warning** — emits a `charts:degraded` Sentry warning when any country is invalid (previously shipped silently under a green run).
- **`fetchPublishedCharts`** — reads the prior payload, degrading to `null` on any failure so carry-forward never aborts the crawl.
- Wire `CHARTS_BLOB_URL` into the cron workflow env (secret already configured).

## Verify

- New unit tests: retry 5, published-charts 5, crawlAll/summarizeValidity 9. Full suite 150/150.
- Local CI matrix green: format, lint, typecheck, test, build.
- Post-merge: `workflow_dispatch` the cron to confirm carry-forward + the degraded path live.

Closes #54

## Summary by CodeRabbit

* **New Features**
  * Chart data carry-forward when current fetch fails for specific regions
  * Enhanced error reporting with validity summaries and degraded state warnings

* **Bug Fixes**
  * Automatic retry mechanism with exponential backoff for transient failures

* **Tests**
  * Added test coverage for retry logic and data recovery scenarios

### CODE REVIEW

**Comment by taewanu** [comment]
<details>
<summary>🤖 Review summary backup — restore from here if GitHub blanks the summary on submit</summary>

_Generated by pr-review-agent v0.0.0 (preview release). Rendering, voice, anchoring, and failure handling are still landing in slices. See open issues for in-progress work._

---

Carry-forward-on-failure for the crawl, with a retry wrapper around Apple RSS and a degraded-signal split in the cron entrypoint.

- Dependency injection keeps every new module testable, and the tests cover the success path plus all failure branches
- Carry-forward edge cases (null previous, missing country, empty prior entry) are each pinned by a test
- Blob path is stable, so reading CHARTS_BLOB_URL back as last-good is sound
- Nothing high-signal to flag

---

🤖 Drafted by [pr-review-agent](https://github.com/taewanu/pr-review-agent). Submit, edit, or cancel as needed.


</details>

**Review by taewanu — COMMENTED** [review]
_Generated by pr-review-agent v0.0.0 (preview release). Rendering, voice, anchoring, and failure handling are still landing in slices. See open issues for in-progress work._

---

Carry-forward-on-failure for the crawl, with a retry wrapper around Apple RSS and a degraded-signal split in the cron entrypoint.

- Dependency injection keeps every new module testable, and the tests cover the success path plus all failure branches
- Carry-forward edge cases (null previous, missing country, empty prior entry) are each pinned by a test
- Blob path is stable, so reading CHARTS_BLOB_URL back as last-good is sound
- Nothing high-signal to flag

---

🤖 Drafted by [pr-review-agent](https://github.com/taewanu/pr-review-agent). Submit, edit, or cancel as needed.

File: scripts/crawl/cron.ts [hunk 1]
```diff
@@ -23,22 +32,43 @@ try {
...
+        Sentry.captureMessage("charts:degraded", {
+          level: "warning",
+          extra: {
+            run_id: process.env.GITHUB_RUN_ID,
+            invalid_codes: summary.invalidCodes,
+            valid_count: summary.validCount,
+            country_count: summary.total,
+          },
+        });
+      }
```
[the remark this answers is not in the document]
Comment (taewanu) [reply scripts/crawl/cron.ts:71]: Intentional divergence — keeping `invalidCodes.length > 0`.

It equals `validCount < total`, so with 40 countries it already is the AC's `valid_count < 40`. The literal `< 40` hardcodes a magic number: if `COUNTRIES` grows (e.g. to 45), `validCount < 40` would stay silent with 4 empty countries (41 ≥ 40) — a mis-fire. The relative check tracks the real set size.

The AC's intent is "warn when the payload ships incomplete," which this captures, and it pairs with carry-forward: the warning fires only when a country fails *and* has no prior good data to fall back on.
